src/learning/gmm.rs

```diff
@@ -209,174 +212,404 @@ impl GaussianMixtureModel {
+    /// let mut gmm: GaussianMixtureModel<Random, CholeskyFull> = GaussianMixtureModel::new(2);
     /// gmm.set_max_iters(5);
     /// ```
     pub fn set_max_iters(&mut self, iters: usize) {
         self.max_iters = iters;
     }
 
-    fn initialize_covariances(&self, inputs: &Matrix<f64>, reg_value: f64) -> LearningResult<Matrix<f64>> {
-        match self.cov_option {
-            CovOption::Diagonal => {
-                let variance = try!(inputs.variance(Axes::Row));
-                Ok(Matrix::from_diag(&variance.data()) * reg_value.sqrt())
-            }
+    // == Parameters
+    // inputs : [n_samples, n_features]
+    //
+    // == Returns
+    // weighted_log_prob : [n_features, n_components]
+    //
+    fn estimate_weighted_log_prob(&self, inputs: &Matrix<f64>) -> Matrix<f64> {
+        let mut log_prob = self.estimate_log_prob(&inputs);
+        // println!("log_prob: \n{:.4}", log_prob.select_rows(&[0, 1, 2, 3, 4, 5, 6]));
+        // println!("mix_weights: \n{:?}", &self.mix_weights);
+        let log_weights = self.mix_weights.iter().map(|w| w.ln());
+        for (lp, lw) in log_prob.iter_mut().zip(log_weights) {
+            *lp += lw;
+        }
+        log_prob
+    }
 
-            CovOption::Full | CovOption::Regularized(_) => {
-                let means = inputs.mean(Axes::Row);
-                let mut cov_mat = Matrix::zeros(inputs.cols(), inputs.cols());
-                for (j, row) in cov_mat.iter_rows_mut().enumerate() {
-                    for (k, elem) in row.iter_mut().enumerate() {
-                        *elem = inputs.iter_rows().map(|r| {
-                            (r[j] - means[j]) * (r[k] - means[k])
-                        }).sum::<f64>();
-                    }
+    // called estimate_log_gaussian_prob in scipy
+    //
+    // == Paramers
+    // inputs : [n_samples, n_features]
+    //
+    //
+    fn estimate_log_prob(&self, inputs: &Matrix<f64>) -> Matrix<f64> {
+        let precisions: &Vec<Matrix<f64>> = self.precisions.as_ref().unwrap();
+        let ref model_means = self.model_means.as_ref().unwrap();
+        // The log of the determinant for each precision matrix
+        let log_det = precisions.iter().map(|m| m.det().ln());
+        // println!("log_det: {:?}", log_det);
+
+        let mut log_prob = Matrix::zeros(inputs.rows(), self.comp_count);
+        for k in 0..self.comp_count {
+            let prec = &precisions[k];
+            // y is a matrix of shape [n_samples, n_features]
+            let mut y = inputs * prec;
+            // println!("y: \n{:.4}", &y.select_rows(&[0, 1, 2, 3, 4, 5, 6]));
+            // Matrix of shape [1, n_features]
+            let z: Matrix<f64> = model_means.select_rows(&[k]) * prec;
+            // println!("z: \n{:.4}", &z);
+            
+            // Subtract the mean of each column from the matrix y
+            for col in 0..y.cols() {
+                for row in 0..y.rows() {
+                    y[[row, col]] -= z[[0, col]];
```

Comment:
Minor stuff here - this access pattern will incur some cache misses.

We should swap the two loops, so that we access contiguous data in sequence (right now we jump down the columns first). We should also probably use `get_unchecked_mut` here, so remove the bound checks that we know we don't need.